shared/src/balancer_event_handler.rs

```diff
@@ -0,0 +1,198 @@
+use crate::{
+    current_block::BlockRetrieving,
+    event_handling::{BlockNumber, EventHandler, EventIndex, EventStoring},
+    impl_event_retrieving,
+    maintenance::Maintaining,
+};
+use anyhow::{anyhow, Context, Result};
+use contracts::{
+    balancer_v2_vault::{
+        self, event_data::PoolRegistered as ContractPoolRegistered, Event as ContractEvent,
+    },
+    BalancerV2Vault,
+};
+use ethcontract::common::DeploymentInformation;
+use ethcontract::{dyns::DynWeb3, Event as EthContractEvent, EventMetadata, H160, H256};
+use std::fmt::{Debug, Formatter};
+use std::ops::RangeInclusive;
+use tokio::sync::Mutex;
+
+#[derive(Debug)]
+pub enum BalancerEvent {
+    PoolRegistered(PoolRegistered),
+}
+
+#[derive(Debug)]
+pub struct PoolRegistered {
+    pub pool_id: H256,
+    pub pool_address: H160,
+    pub specialization: PoolSpecialization,
+}
+
+/// There are three specialization settings for Pools, which allow for cheaper swaps at the cost of reduced
+/// functionality:
+///
+///  - General: no specialization, suited for all Pools. IGeneralPool is used for swap request callbacks, passing the
+/// balance of all tokens in the Pool. These Pools have the largest swap costs (because of the extra storage reads),
+/// which increase with the number of registered tokens.
+///
+///  - Minimal Swap Info: IMinimalSwapInfoPool is used instead of IGeneralPool, which saves gas by only passing the
+/// balance of the two tokens involved in the swap. This is suitable for some pricing algorithms, like the weighted
+/// constant product one popularized by Balancer V1. Swap costs are smaller compared to general Pools, and are
+/// independent of the number of registered tokens.
+///
+///  - Two Token: only allows two tokens to be registered. This achieves the lowest possible swap gas cost. Like
+/// minimal swap info Pools, these are called via IMinimalSwapInfoPool.
+#[repr(u8)]
+pub enum PoolSpecialization {
+    General = 0,
+    MinimalSwapInfo = 1,
+    TwoToken = 2,
```

Comment:
Maybe include a link to the Balancer docs? Also, I would suggest having the different pool speiciliazation docs inline:
```suggestion
/// minimal swap info Pools, these are called via IMinimalSwapInfoPool.
#[repr(u8)]
pub enum PoolSpecialization {
    /// No specialization, suited for all Pools. ...
    General = 0,
    /// IMinimalSwapInfoPool is used instead ...
    MinimalSwapInfo = 1,
    /// Only allows two tokens to be registered. ...
    TwoToken = 2,
```

Reply:
> Maybe include a link to the Balancer docs?

On second thought... I don't think this has much value.